### PR TITLE
[mlir][vector] Add support for distributing masked writes

### PR DESCRIPTION
**Pull request by qedawkins**
Because the mask applies to the un-permuted write vector, we can simply distribute the mask identically to the vector, if present.

### CODE REVIEW

**Review by antiagainst — CHANGES_REQUESTED** [review]
Hmm, I'm not sure actually this is correct. In my mental model about the masking, it's applied _close_ to the memory, immediately before/after write/read, because that's where masking really happens. So for write, the masking should match with the memory shape, not exactly the vector shape, when we have non-idenity permutation maps. I think we need inverse the map and compose with the vector shape to get the mask shape, and distribute that.. @dcaballe WDYT?

**Comment by qedawkins** [comment]
> Hmm, I'm not sure actually this is correct. In my mental model about the masking, it's applied _close_ to the memory, immediately before/after write/read, because that's where masking really happens. So for write, the masking should match with the memory shape, not exactly the vector shape, when we have non-idenity permutation maps. I think we need inverse the map and compose with the vector shape to get the mask shape, and distribute that.. @dcaballe WDYT?

Sorry, I reread the docs for transfer_write and you're right. Moving to a draft for now.

**Comment by antiagainst** [comment]
Actually given we are only handling 1-D distribution in transfer write at the moment, and the fact we are [compressing unused dims when `inferTransferOpMaskType`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L3758-L3773), effectively that guarantees we have the same shape as the vector type? I'm a bit confused now; better to get @dcaballe to take another look.

We can enable support step-by-step. The current impl is fine if we additionally check the transfer write has an identity map I think.

**Comment by antiagainst** [comment]
The doc in transfer write w.r.t. masking is confusing--[this part](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L1458-L1460) seems obsolete and need to be deleted. I'll send a pull requst to fix it.

**Comment by antiagainst** [comment]
Uploaded https://github.com/llvm/llvm-project/pull/71490 for it.

**Comment by qedawkins** [comment]
> Actually given we are only handling 1-D distribution in transfer write at the moment, and the fact we are [compressing unused dims when `inferTransferOpMaskType`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L3758-L3773), effectively that guarantees we have the same shape as the vector type? I'm a bit confused now; better to get @dcaballe to take another look.
> 
> We can enable support step-by-step. The current impl is fine if we additionally check the transfer write has an identity map I think.

yeah after further thought, n-D masked writes will be quite tricky to handle properly because the warp op does not properly capture the distribution mapping needed by the transfer_write. We would have to do something like, apply the transfer map, distribute, and reverse the map to get back to the original mask shape. Ideally those transposes would cancel, but step-by-step sounds good to me. I'll guard a bit better here, but a discussion about how to think about these masks would be great.

**Comment by dcaballe** [comment]
Basically this:

> it's applied close to the memory, 

We mask the memory access in whatever shape happens on memory.

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp [hunk 1]
```diff
@@ -489,10 +499,26 @@ struct WarpOpTransferWrite : public OpRewritePattern<vector::TransferWriteOp> {
...
+      // TODO: Distribution of masked writes with non-trivial permutation maps
+      // requires the distribution of the mask to elementwise match the
+      // distribution of the permuted written vector. Currently the details
+      // of which lane is responsible for which element is captured strictly
+      // by shape information on the warp op, and thus requires materializing
+      // the permutation in IR.
+      if (!writeOp.getPermutationMap().isMinorIdentity())
+        return failure();
+      maskType =
+          getDistributedType(writeOp.getMask().getType().cast<VectorType>(),
```
Comment (antiagainst) [inline mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:514]: Nit: IIRC we have getMaskType directly?

**Review by antiagainst — APPROVED** [review]

**Pull request merged**